alibi/explainers/backend/tensorflow/counterfactuals.py

```diff
@@ -0,0 +1,734 @@
+    interface description and `alibi.explainers.backend.tensorflow.counterfactuals` `get_numerical_gradients` method for \
+    an integration example. The kwargs of this function should be specified in the ``'kwargs'`` field
+
+    - ``'grad_fcn'`` should be a callable taking the same inputs as ``'fcn'``. Represents the gradient of the prediction \
+    term wrt model output. The output of this function is multiplied by the numerical gradient to give the gradient of \
+    the non-diferentiable wrt to the input
+   
+    - ``'grad_fcn_kwargs'`` specify the kwargs of ``'grad_fcn'``. 
+"""
+
+
+def range_constraint(
+        X: Union[tf.Variable, tf.Tensor],
+        low: Union[float, np.ndarray],
+        high: Union[float, np.ndarray]) -> Union[tf.Variable, tf.Tensor]:
+    """
+    Clips `X` so that the its values lie within the :math:`[low, high]` interval. This function is used to constrain the
+    values of the counterfactuals with the specified range.
+
+    Parameters
+    ----------
+    X
+        Tensor or Variable to be constrained.
+    low, high
+        If the input is of type `np.ndarray`, then its shape has to be such that it supports element-wise comparison
+        with `X` is possible.
+
+    Returns
+    -------
+    A tensor whose values are constrained.
+
+    Raises
+    ------
+    ValueError:
+        If `low` > `high`.
+    """
+
+    feat_interval = high - low
+    if np.any(feat_interval < 0.):
+        raise ValueError(
+            f"Invalid constrain. Lower bound needs to be smaller upper bound but the difference was "
+            f"{feat_interval} for an element.",
+        )
+
+    return tf.clip_by_value(X, clip_value_min=low, clip_value_max=high)
+
+
+def slice_prediction(prediction: tf.Tensor, target_idx: Union[int, str], src_idx: int) -> tf.Tensor:
+    """
+    Returns a slice from `prediction` depending on the value of the `target_idx`.
+
+    Parameters
+    ----------
+    prediction
+        A prediction tensor
+    target_idx
+
+        - If ``int``, the slice of `prediction` indicated by it is returned
+        - If 'same', the slice indicated by `src_idx` is returned
+        - If 'other', the slice that maximises the prediction other than `src_idx` is returned
+
+    src_idx
+        An index in the prediction that influences the returned slice as described above.
+    """
+
+    if isinstance(target_idx, int) or target_idx == 'same':
+        return prediction[:, target_idx]
+    if target_idx != 'other':
+        raise ValueError("The only allowed values for output slices are {'same','other'}")
+
+    _, indices = tf.math.top_k(prediction, k=2)
+    if indices[0][0] == src_idx:
+        cf_prediction = prediction[:, indices[0][1].numpy()]
+    else:
+        cf_prediction = prediction[:, indices[0][0].numpy()]
+    return cf_prediction
+
+
+def wachter_blackbox_wrapper(X: Union[tf.Tensor, tf.Variable],
```

Comment:
Yes, only used for black box. This is so that I could comply with the interface for the numerical gradient function and keep that one generic. 

This might be more generic than the name suggests but different functions are used in the proto implementation so didn't see a point to rename it.